app/services/auto_end_previous_reservation.rb

```diff
@@ -0,0 +1,47 @@
+# frozen_string_literal: true
+
+class AutoEndPreviousReservation
+
+  def initialize(product, current_user)
+    @product = product
+    @current_user = current_user
+  end
+
+  def end_previous_reservations!
+    return unless SettingsHelper.feature_on?(:auto_end_reservations_on_next_start)
+    return unless timer_based_instrument?
+
+    reservations_to_end = previous_reservations
+
+    reservations_to_end.each do |reservation|
+      auto_end_reservation(reservation)
+    end
+  end
+
+  private
+
+  def timer_based_instrument?
+    @product.control_mechanism == Relay::CONTROL_MECHANISMS[:timer] ||
+      @product.control_mechanism == Relay::CONTROL_MECHANISMS[:relay]
+  end
+
+  def previous_reservations
+    @product.reservations.joins(:order_detail)
+            .where(actual_end_at: nil)
+            .where(reserve_end_at: ...Time.current)
+            .where(order_details: { canceled_at: nil })
+            .where.not(actual_start_at: nil)
```

Comment:
I  think we should limit to the most recent reservation and define a timewindow for reservations that can be considered.

A few questions that come to my mind:
- Does it make sense for this to end a reservation if it was several days ago? what about one day ago? what's the limit? 
- Does it make sense to end several reservations and setting all of them the same actual end time?

Reply:
we already have a task called AutoExpireReservation that runs every 5 minutes and look for reservations with more than 12 hs and move them to the problem queue, so there aren't many active ones hanging around. This feature is complementary, as it handles the case when someone arrives before those 12 hours.

Multiple active reservations per instrument are rare due to validations, but if it happens (mostly admin reservations), ending them all with the same timestamp makes sense since that's when the new user actually took control.

Do you think I should add a time limit anyway?